### ISSUE TITLE
Question about RoPE Implementation in modeling_llama: Should torch.cat be repeat_interleave?

### ISSUE THREAD

**Issue by abhidipbhattacharyya** [huggingface/transformers#40462]
Hi,
I was going through the code for `modeling_llama` and the RoPE implementation. I came across the following function:

```
def forward(self, x, position_ids):
        inv_freq_expanded = self.inv_freq[None, :, None].float().expand(position_ids.shape[0], -1, 1).to(x.device)
        position_ids_expanded = position_ids[:, None, :].float()

        device_type = x.device.type if isinstance(x.device.type, str) and x.device.type != "mps" else "cpu"
        with torch.autocast(device_type=device_type, enabled=False):  # Force float32
            freqs = (inv_freq_expanded.float() @ position_ids_expanded.float()).transpose(1, 2)
            emb = torch.cat((freqs, freqs), dim=-1)
            cos = emb.cos() * self.attention_scaling
            sin = emb.sin() * self.attention_scaling

        return cos.to(dtype=x.dtype), sin.to(dtype=x.dtype)

```
I believe the line `emb = torch.cat((freqs, freqs), dim=-1)` should be replaced with `repeat_interleave`. This is because the cosine/sine angles for matrix multiplication should be structured like:
```
[cos(θ₁), cos(θ₁), cos(θ₂), cos(θ₂), cos(θ₃), cos(θ₃), ...]

```
This way, further down the stream when we compute:
```
q_embed = (q * cos) + (rotate_half(q) * sin)
```
...the values are aligned properly for pairwise rotation. However, the current `torch.cat((freqs, freqs), dim=-1) ` should produce:
```
[cos(θ₁), cos(θ₂), cos(θ₃), cos(θ₁), cos(θ₂), cos(θ₃), ...]
```
which seems incorrect. Am I missing something?
Thanks,
Abhidip

**Comment by vasqu** [comment]
It depends on the rotation on the weights as well, with llama we have a half half split - see the rotation here https://github.com/huggingface/transformers/blob/c8c7623f20eb1f4bef5ba78507e7a55406977a95/src/transformers/models/llama/modeling_llama.py#L109-L112

There are models that use your rotation instead and then they repeat interleave, e.g. glm see https://github.com/huggingface/transformers/blob/c8c7623f20eb1f4bef5ba78507e7a55406977a95/src/transformers/models/glm/modeling_glm.py#L103-L135

You can exchange them if you reshape the weights correctly ;)

**Comment by vasqu** [comment]
Closing this (assuming it's answered by the reaction), but reopen this if not 🤗
